### ISSUE TITLE
security.ssl3.*

### ISSUE THREAD

**Issue by earthlng** [ghacksuserjs/ghacks-user.js#24]
## proposals
* [x] A: remove 1214 period, I'm sorry but I trust the mozilla specialists more on this than @pyllyukko 
  * Roman says: I am open to making all of 1214 commented out and adding more to the description, but will not remove.
  * 1214 needs renaming
  * earthling says: renamed it, but wontchange atm so we match @pyllyukko's user.js
* [x] B: split 1210 into 1210 and 1210a or something like that.
  * Roman says: No problems with splitting it
  * earthling says: wontchange atm so we match @pyllyukko's user.js
* [x] C: add the 2 static-rsa prefs, but definitely as inactive
  * Roman says: yes, what do we call it?
  * earthling says: pull request pending, no further change needed

## original post
I looked at our current settings and I want to share some of my observations.

https://badssl.com/
https://www.howsmyssl.com/
https://wiki.mozilla.org/Security/Server_Side_TLS
https://www.eff.org/deeplinks/2015/10/how-to-protect-yourself-from-nsa-attacks-1024-bit-DH

https://dh2048.badssl.com/ => we break dh2048 due to `user_pref("security.ssl3.dhe_rsa_aes_256_sha", false);`
The eff article explicitly names that pref and recommends to disable it even though it talks about 1024-bit primes being presumably broken by the NSA, and that pref uses a 2048-bit prime.
There's also this sentence from the mozilla wiki for "Modern compatibility":
- DHE is removed entirely because it is slow in comparison with ECDHE, and all modern clients support elliptic curve key exchanges.

It doesn't say it's easy to break only that it's slow. Do we assume the NSA can also break DHE with a 2048-bit prime, or should we comment out that pref for less breakage (and recommend it only for the paranoid?) - because according to badssl.com dh2048 is still secure.
If we keep it, we may want to rename 1210 to 'disable DHE (Diffie-Hellman Key Exchange)' and remove the 1024 bit part.

'Modern compatibility' ciphers and the matching pref:
01. ECDHE-ECDSA-AES256-GCM-SHA384  => security.ssl3.ecdhe_ecdsa_aes_256_gcm_sha384
02. ECDHE-RSA-AES256-GCM-SHA384      => security.ssl3.ecdhe_rsa_aes_256_gcm_sha384
03. ECDHE-ECDSA-CHACHA20-POLY1305 => security.ssl3.ecdhe_ecdsa_chacha20_poly1305_sha256
04. ECDHE-RSA-CHACHA20-POLY1305	=> security.ssl3.ecdhe_rsa_chacha20_poly1305_sha256
05. ECDHE-ECDSA-AES128-GCM-SHA256 => security.ssl3.ecdhe_ecdsa_aes_128_gcm_sha256
06. ECDHE-RSA-AES128-GCM-SHA256 => security.ssl3.ecdhe_rsa_aes_128_gcm_sha256
07. ECDHE-ECDSA-AES256-SHA384 => security.ssl3.ecdhe_ecdsa_aes_256_sha
08. ECDHE-RSA-AES256-SHA384 => security.ssl3.ecdhe_rsa_aes_256_sha
09. ECDHE-ECDSA-AES128-SHA256 => security.ssl3.ecdhe_ecdsa_aes_128_sha (disabled by 1214)
10. ECDHE-RSA-AES128-SHA256 => security.ssl3.ecdhe_rsa_aes_128_sha (disabled by 1214)

Interestingly Firefox sends them in a different order (see howsmyssl.com).
Instead of 1-8 its **5,6,3,4,1,2**,7,8
When 9+10 are enabled the order is 5,6,3,4,1,2,**7,9,8,10**
(+ `TLS_RSA_WITH_AES_128_CBC_SHA` and `TLS_RSA_WITH_AES_256_CBC_SHA` at the end [11+12], for the 2 remaining ones that we still allow)

The Rationales explain some of it:
In **Modern compatibility**:
_AES256-GCM is prioritized above its 128 bits variant, and ChaCha20 because we assume that most modern devices support AESNI instructions and thus benefit from fast and constant time AES._

While in **Intermediate compatibility (default)**:
_ChaCha20 is prefered as the fastest and safest in-software cipher, followed by AES128. Unlike the modern configuration, we do not assume clients support AESNI **and thus do not prioritize AES256 above 128 and ChaCha20**. There has been discussions on whether AES256 extra security was worth its computing cost in software (without AESNI), and the results are far from obvious. At the moment, AES128 is preferred, because it provides good security, is really fast, and seems to be more resistant to timing attacks._

Well, anyhow, there's nothing we can do about it anyway.

Apart from the 10 modern ciphers (2 of which we have disabled) we only support 2 older ones: `security.ssl3.rsa_aes_128_sha` and `security.ssl3.rsa_aes_256_sha`, which is why `static-rsa` (https://static-rsa.badssl.com/) still works, but it's listed with an `!` and not a red warning sign.

The only other thing that I really don't like is that https://incomplete-chain.badssl.com/ still works.
But maybe this can be fixed if `security.nocertdb` is set to true; I haven't tested this because I really don't like the current implementation behind that pref.

IMO we should remove 1214, because the 128bit in AES is not the problem, 5+6 have them also.
And the 2 prefs in 1214 are part of the 10 modern ciphers.
Perhaps we were a bit overeager when we added 1214, without really understanding it. (?)

**Comment by earthlng** [comment]
What bugs me a bit is the "Restore default settings" button on those "Secure Connection Failed" pages.
I want to research what that button exactly does, and if there's a pref to hide/remove that button altogether.

**Comment by Roman-Nopantski** [comment]
great work there :) ciphers are not my domain. some thoughts:

- 1210: `security.ssl3.dhe_rsa_aes_256_sha"`
  * Why enable it? It's not in your list of 1-10, or was it not included?
  * Site breakage? Yet to see any. I doubt its hardly ever used and is fast becoming obsolete
  * rename: definitely expand the acronym
  * choice 1: prefer: keep both prefs there and both un-commented out. No changes except change name as per your suggestion (there's no more DHE, right?)
  * choice 2: messy: If we remove it then your name suggestion is not right as it doesn't cover *all* DHE, but will still be alright if we merely comment it out. Or we could move it to a different pref number of its own, in which case neither name change works .. and we can't remove it from the js entirely because we should keep it there to track/remember it and for end users benefit.

- 1214: 10 modern ciphers (2 of which we have disabled)
  * interesting order, 9 preferred before 8
  * I have yet to feel any site breakage from these, surely if they are 128bit (and that clears up the low bit side of things) and are, at best, the 8th preferred cipher by FF, then I think we should still block using them

- we support 2 older ciphers
  * these are: `security.ssl3.rsa_aes_128_sha` and `security.ssl3.rsa_aes_256_sha`
  * why is 1214 titled disable 128 when the second has 256 in it's name?
  * these give an exclamation mark on that test, so clearly they are on their way out
  * what would site breakage be? Little is my guess: yet to see anything myself or get complaints
  * what does pyllyukko say (I am, loath to @ him, he has us on watch anyway)

===
- I don't like  security.nocertdb -> true : it removes functionality - all logins & pw's as well as CAs are session only, so basically its the same as disabling password saving plus the CA thing. In a super hardened js it might make sense. One day we'll branch a super-hardened version.

- restore default settings button on secure connection fail: I actually used that the other day if you remember our recent ghacks discussion, worked a treat (I compared a live prefs.js with a copy and spotted the changes in real time) - that's how I found those two ciphers. There is no way I think we should stop end users being able to reset these. Don't forget, the user.js will override them.

**Comment by Atavic** [comment]
I have set these to false:
```
security.ssl3.ecdh_ecdsa_rc4_128_sha
security.ssl3.ecdh_rsa_rc4_128_sha
security.ssl3.ecdhe_ecdsa_rc4_128_sha
security.ssl3.ecdhe_rsa_rc4_128_sha
```
Unsecure cryptographic keys aren't issued anymore. So there's a limbo where you stumble on expired CA Certs until they are renewed (and use different keys).

I abstain from crosslinking the discussions on [pyllyukko's user.js](https://github.com/pyllyukko/user.js/issues) repo.

**Comment by earthlng** [comment]
Thank you ;)
1210 - no, it's not included because we have it set to false. 1-10 are all the 'modern' ciphers only.
_there's no more DHE, right?_ - if I understand you correct: no, the 2 prefs in 1210 are the only old-school DHE, ie without 'elliptic curves'. But all the modern ones still use Diffie-Hellmann but **with** EC hence ECDHE.
I just brought it up because badssl.com considers the 2nd pref (2048bit) still secure and since you want to keep a balance and low-impact and everything. Disabling it increases the JS-less fingerprint on HTTPS sites blabla. I'll create a pull-request and we can go from there.

1214 - `surely if they are 128bit` - but that's the thing 5+6, ie the 1st and 2nd choice in FF atm, also use 128AES. They add in some gcm (hmmm sounds delicious, whatever that is) but they ARE the preferred/best ciphers atm. so it can't be that bad.
They are part of the 10 modern ciphers, I'd say we remove 1214.
I saw them originally in @pyllyukko user.js and we also added them because of that.
And I have no problem admitting that I didn't really understand it back then. I only saw 128bit ergo bad.

`why is 1214 titled disable 128 when the second has 256 in it's name?` it doesn't - you mixed that up buddy ;) or do you mean in the 1-10 list? AES128-SHA256

`we support 2 older ciphers` those 2 are not in the user.js. member when that security site broke for you and you had to figure out why? oh i member - those 2 ;) badssl has them as 'static-rsa'

'security.nocertdb' I think they are still trying to figure out how to achieve disabling only the certdb without also breaking every other db. Until we have a hardened, chances are they improve it at least slightly.

'restore default settings button' - i didn't dare click on it tbh. only want to see what exactly it does and if it can be blocked. purely out of interest. I can hide/disable the button with custom css on my own machine, but I wanna see if there's a pref for that. (fe. for company environment or so)

@Atavic those are all the RC4 ciphers and they were removed recently from FF anyway.

**Comment by Roman-Nopantski** [comment]
- 1210: "it's not included because we have it set to false" - well I thought you might have enabled it for the test. Where would it come in the prefered order if enabled?
- 1214: `titles` came copypasted from pyllyukko js but of course ours are bit pieces and I probably screwed it up. What should we describe 1214 as then?
- 128bit: its all over my head, but we both know the bit is not the only component in determining the a cipher strength
- `we support 2 older ciphers` - I know they're not in the user.js because they are at default and there is no need to change them.
- "want to keep a balance and low-impact and everything" - but there is virtually no breakage with this

==OT ==
I think I need to get the readme finished and outline some preferred stances for defaults (users can customize). I'm not too worried about site breakage as much as not putting "newbies" at risk - I need some lowest common denominator, but will expect some level basic knowledge and common sense. I want this to be as easy as possible for newcomers who may struggle. To that purpose I am looking at turning on TP and SB and app+extension update checks (see #10 ). TP + SB can work in tandem and means anyone just running it and they don't have uBo for example, won't be at risk. The app update check is sound advice to make sure users are reminded their is a new release - not all users like those of us currently here are avid (rabid) checkers of everything FF related. These changes would be what I call "best practice", and that's what I need to outline.

Its difficult. There is no general rule, and weight has to be given to each variable. So simply saying security trumps privacy is never always going to hold true, and gets even worse when we match those against site breakage. Would we be prepared to (by default) break youtube+facebook+twitter simply to block some privacy or fingerprinting issue - no. Would we do the same if only a handful of obscure sites broke - yes. Same with allowing history by default and not starting in PB mode, and not auto-wiping cookies: too much functionality loss in the default js will just drive people away. That's part of my/our end game here - I see an opportunity to get this mainstream - hence I'm "dumbing" it down so to speak, so out of the box its as secure/private as we can get with as little breakage/incovenience (which it already is bar a couple of items) - and we'll never please everyone, but we have given them the means and knowledge to tweak, which is another goal.

==back to the topic==
So what exactly are you proposing to change. Just bullet point it without any descriptions eg 1. remove X 2. comment out Y , because I'm confused now, which is not hard to do. :)

**Comment by earthlng** [comment]
_Where would it come in the prefered order if enabled?_ - after 10 and before the 2 static-rsa ones

1. remove 1214 period, I'm sorry but I trust the mozilla specialists more on this than @pyllyukko 
2. maybe split up 1210 into 1210 and 1210a or something like that. I can work on a pull-request for this.
3. perhaps add the 2 static-rsa prefs, but definitely as inactive (commented out), for people who wanna go full 'modern' and who can deal with breakage themselves. It broke that security site for you recently, so some, maybe a lot, of sites still use that old stuff if they need/want to support older devices and OSes.

OT: I think you go waaaay overboard with wanting to make this more mainstream.
IMHO most people don't care to read all the stuff we collect for them, no matter how much you dumb it down. "where do i put this" etc

**Comment by Roman-Nopantski** [comment]
- adding those two static-rsa's as commented out, for experts, run at your own risk, with warnings - I am ok with, so go ahead and patch that

OT: No I think you misunderstand - I think we are basically already there bar a few of things (SB, TP, update checks) which have the potential to really put end users in a bad spot. The problem is you went overboard in my interpretation :) Now chill out or I'll sik Rex into ya

**Comment by earthlng** [comment]
OT: yeeeah, no. So what about the whole Font section fe - you think they'll appreciate that?
or the webgl stuff .. "you broke my super-cool shit-farming browser game"
"why can I not watch youporn vids fullscreen anymore, you cunt?"
"all my videos of super funny cats lag now, you cunt!" (hw-accel)
"where did all my 700 open tabs of funny cat videos from yesterday go?" .... "you cunt!"
and I'm just scrolling thru the file and picking out a few of the 400+ prefs right now.
Do I need to say more? xD

**Comment by Roman-Nopantski** [comment]
Just getting this off my chest. Am not mad at you by any means bro, but for god's sake, let me the damn thing finished before you poke ridiculous holes in it.

Fonts do not BREAK SITES. So what if we clear history on close, that does not BREAK SITES. Those are minor inconveniences. Full-screen is not disabled and if it was its under the PERSONAL section. The three items I *suggested* and have up for *discussion* (SB + TP + some update checks) I gave solid reasons for them - they too do not BREAK SITES. nor do they compromise privacy, and they add security and privacy, which is important. You're being a bit pedantic. Everyone will have different needs and different views. I already said the js is pretty much spot on, but you want to keep arguing a point.

security vs privacy vs fingerprinting vs functionality (for both app and websites  which we can treat differently) vs inconvenience (app and websites) vs popularity or how widespread the damage is vs a dozen other factors if you really wanted to science the sh*t out of it.

If I said I want things generally in this order
 > security trumps privacy which trumps fingerprinting and the effects they cause is weighed against (high to low) site breakage, then inconvenience

If we followed that for everything, the browser would be hardened. We would be starting in PB mode, disabling passwords, not have history etc because they are only inconveniences. We wouldn't even have cookies like pyllyukko's does, and his is *hardened*. The thing is the weighing will vary for each pref on its own merits.

We have over 270 numbered items, and its been added to adhoc for 6 months on my own and then 18 months at ghacks, basically with what I wanted (remember all those "this is the authors settings choose your own" comments etc). Now it is not about ME and what was a select few commenters at ghacks. It's about properly setting up a purpose, defining some sort of best practice, and reviewing what we have. And every single item needs addressing on its own merits. There is no real ONE RULE TO RULE THEM ALL. **Scanning thru the list and pointing out inevitable paradoxes is ridiculous** , especially when I haven;t even had a chance finish our readme purpose, or to to review and bring up for discussion anything glaring out of place. You know there are trade-offs.

Most of what we have is spot on bar those three things I brought up (but I have yet to a review) which I think can be changed for very good reasons. But we're also not going full-on super-hardened, so eg we won't do start in PB mode (privacy) because we want some convenience (history) but we will temper than with clearing history on close. *See, everything is a compromise*.

I have already thought about videos, and as long as youtube works for the masses I'm not too concerned. Service Workers section would be the only section I'm a bit iffy about. And we can always review - which is what I am doing when I propose something.

And the last point I will make is that this is just the default user.js, and users can customize it. Experienced and knowledgeable users are more likely to diverge than "newbies", so that in itself is a best practice, so that there is as little unintended *damage* as possible - do not confuse that with *site breakage*. That's just being responsible.

**Comment by pyllyukko** [comment]
> what does pyllyukko say (I am, loath to @ him, he has us on watch anyway)

Huh? There's so much going on here :)

Here's the reason to (slowly) move away from 128-bit key sizes:
* <https://www.iad.gov/iad/library/ia-guidance/ia-solutions-for-classified/algorithm-guidance/cnsa-suite-and-quantum-computing-faq.cfm>
* [Grover's algorithm](https://en.wikipedia.org/wiki/Grover's_algorithm)

**Comment by Roman-Nopantski** [comment]
@earthlng FYI: removed all OT, added the three proposals as a check list to your first post. As pyllyukko says "so much going on here". Which one the three proposals does pullyukko's "reason to (slowly) move away from 128" relate to?

**Comment by earthlng** [comment]
Thanks Pantski, good stuff. 
I read your last OT before you removed it - duly noted. Let's move on. (I'm OT8 now, btw ;)

@pyllyukko well, the 1st site gives me a SEC_ERROR_UNKNOWN_ISSUER and the 2nd one, oh boy the 2nd one.... well let's just say it's slightly over my head (lol, what an understatement) but what matters, if I understood anything it says, is the 2**64 vs 2**128 part. I did some calculations and if the fastest super-computer would suddenly be able to do quantum-computing today it would still take ~92secs of full-capacity computing time to brute-force AES128. I'm not a criminal so IMHO even those 92secs would still not be worth it to get to my encrypted data. Imagine what such a machine could do in those 92 seconds other than brute-forcing my encrypted data! And with TLS1.3 and if session-tickets are disabled, afaik there's no longer a way to store everything and crack it later, meaning it's full forward-secrecy.

All that aside, that's not really the point here. Why do we disable 2 of the 10 modern ciphers because of a theoretical future possibility but at the same time allow 2 older ciphers for compatibility reasons, with one of them also using AES128, and both being arguably weaker than the 2 modern ones?

**Comment by earthlng** [comment]
Oh, and there's also a theory that says that AES128 is potentially more secure than AES256.
I mean, idk, it's all very theoretical at the moment anyway. Let's deal with the here and now.

**Comment by Atavic** [comment]
1214 can be commented out or left as is but Im' against its removal. A [link](https://www.ietf.org/mail-archive/web/tls/current/msg20102.html).

**Comment by earthlng** [comment]
1214: won't be removed completely. But whatever we end up doing with it, we should agree on something and then use it both here and @pyllyukko user.js. That way we will at least increase "our little group" slightly. (re: cipher fingerprint)

1209: is already set to 4 in https://dxr.mozilla.org/mozilla-central/source/security/manager/ssl/security-prefs.js#6 so I think this will be default in FF52, idk. It can break sites on servers that haven't updated their openssl or whatnot, hence I think we should wait for FF52, at least. Users can enable it already if they want.

**Comment by earthlng** [comment]
from Atavic's first link:

> Now, I rather think that we should be moving TLS to use Quantum Resistant cryptography; but as of right now, TLS is rather far from that goal, and the symmetric key size is a minor issue;

therefore I'd advocate for 1214 to be commented out, because they are 2 of the 10 modern ciphers

**Comment by earthlng** [comment]
Unfortunately the security-prefs.js also still has the following prefs, so cipher FP will remain a slight problem for us, for now.
```
pref("security.ssl3.dhe_rsa_aes_128_sha", true);
pref("security.ssl3.dhe_rsa_aes_256_sha", true);
pref("security.ssl3.rsa_des_ede3_sha", true);
```

What if we regroup parts of that section, and add a note that changing the default values for them can be fingerprinted? for example:

`12xx: disable insecure ciphers (Warning: changing the values from FF's defaults, can be used for Fingerprinting)`
-> and then move all the relevant prefs into that sub-group, atm 1210, 1213, 1214, each as 12xx a, b, c.

**Comment by Roman-Nopantski** [comment]
Just to be clear: I don't really care about cipher fingerprints. You will not defeat fingerprinting this way - there are just too many holes, even we patch/close dozens of them, there are still too many left. You beat it thru OpSec.

**Comment by earthlng** [comment]
> What bugs me a bit is the "Restore default settings" button on those "Secure Connection Failed" pages.
> I want to research what that button exactly does

It resets all user-set prefs for `security.tls.version.*` and `security.ssl3.*`

https://dxr.mozilla.org/mozilla-central/source/browser/base/content/browser.js#2991
https://dxr.mozilla.org/mozilla-central/source/browser/base/content/browser.js#2910

**Comment by Atavic** [comment]
It does the same as _reset_ on an about:config entry. You can right-click on an about:config entry and there's _reset_ option in the GUI list.

**Comment by earthlng** [comment]
* Roman-Nopantski changed the title from security.ssl3.* to ToDo: security.ssl3.* 4 hours ago
* Roman-Nopantski added the task label 4 hours ago
* earthlng was assigned by Roman-Nopantski 4 hours ago

why all those changes on a closed issue? Do you want me to do something here?
As far as I'm concerned if pyllyukko and the boys are not willing to change these, neither should we.

**Comment by Roman-Nopantski** [comment]
Just standardizing the issue titles and lables - that original title was from 15 days ago when we just started .. since it had three proposals I made it a ToDo and added some labels. When I do a write up for Martin I want to be able to say we did these enhancements (foldable formating, tags, revamped sections, a wiki, etc). I could just as easily change the title to discussion: blah and change labels again.
